### ISSUE TITLE
Add Browser-Based SAM (Segment Anything Model) Support via ONNX Runtime Web

### ISSUE THREAD

**Issue by hughlv** [bitroc-ai/annota#12]
# Add Browser-Based SAM (Segment Anything Model) Support to Annota

## Summary

Add native browser-based segmentation using Meta's Segment Anything Model (SAM) via ONNX Runtime Web. This would provide a new interactive tool that allows users to click on objects and instantly generate polygon annotations without requiring any backend API.

## Motivation

**Current Limitations:**
- Interactive segmentation currently requires backend API calls
- Network latency affects user experience
- Requires GPU server infrastructure
- Privacy concerns with sending images to external services
- Offline usage not possible

**Benefits of ONNX-based SAM:**
- ✅ **Zero latency** - Runs entirely in browser
- ✅ **No backend required** - Pure client-side inference
- ✅ **Privacy-first** - Images never leave the device
- ✅ **Offline support** - Works without internet
- ✅ **Cost-effective** - No GPU server costs
- ✅ **Framework agnostic** - Works with any image viewer

## Reference Implementation

Meta AI's official SAM demo demonstrates this approach:
- **Repository**: https://github.com/facebookresearch/segment-anything/tree/main/demo
- **Live Demo**: https://segment-anything.com/demo
- **Key Implementation**: Uses ONNX Runtime Web with WebAssembly for browser-based inference

### Architecture Overview

**Two-Stage Pipeline:**

1. **Image Encoder** (Offline/Cached) - Generates image embeddings
   - Model: `sam_vit_h_encoder.onnx` (~350MB) or pre-computed `.npy` files
   - Input: Image (1024×1024)
   - Output: Embedding tensor [1, 256, 64, 64]
   - **Execution**: Pre-compute offline and cache, or encode on-the-fly

2. **Mask Decoder** (Interactive/Real-time) - Generates masks from prompts
   - Model: `sam_vit_h_decoder.onnx` (~10MB)
   - Input: Image embedding + click coordinates + labels
   - Output: Binary mask [1, 1, 256, 256]
   - **Execution**: Fast inference on every user interaction

#### Phase 1: Decoder-Only (Recommended Starting Point)

**Approach**: Use pre-computed image embeddings
- Model size: ~10MB (decoder only)
- User provides: Pre-computed `.npy` embedding files
- Inference speed: ~100-300ms per click (WebAssembly)
- Memory usage: ~500MB

**Workflow:**
```
1. Load decoder model on tool activation
2. Load precomputed embedding for current image
3. User clicks on object
4. Run ONNX inference with click coordinates
5. Convert mask tensor to binary PNG
6. Use existing loadMaskPolygons() to create annotation
```

#### Phase 2: Full Browser Encoding (Future Enhancement)

**Approach**: Real-time image encoding in browser
- Model size: ~360MB (encoder + decoder)
- Inference speed: ~2-5s for encoding, ~100-300ms per click
- Memory usage: ~2GB
- Enables dynamic images without preprocessing

### Dependencies

```json
{
  "onnxruntime-web": "^1.14.0"
}
```

**Optional for pre-computed embeddings:**
```json
{
  "npyjs": "^0.4.0"
}
```

#### 1. **Model Initialization** (from `App.tsx`)

```typescript
import * as ort from 'onnxruntime-web';

// Configure ONNX Runtime for WebAssembly
ort.env.wasm.numThreads = 1;
ort.env.wasm.simd = true;

// Load decoder model
const model = await ort.InferenceSession.create('/models/sam_decoder.onnx', {
  executionProviders: ['wasm'],
  graphOptimizationLevel: 'all'
});
```

#### 2. **Input Tensor Preparation** (from `onnxModelAPI.tsx`)

```typescript
// Point coordinates: [1, num_points, 2]
const pointCoords = new ort.Tensor('float32',
  new Float32Array([
    clickX * modelScale, clickY * modelScale,  // Actual click
    0, 0                                         // Padding point
  ]),
  [1, 2, 2]
);

// Point labels: [1, num_points]
const pointLabels = new ort.Tensor('float32',
  new Float32Array([1, -1]),  // 1=foreground, -1=padding
  [1, 2]
);

// Mask input: [1, 1, 256, 256] (no prior mask)
const maskInput = new ort.Tensor('float32',
  new Float32Array(256 * 256).fill(0),
  [1, 1, 256, 256]
);

// Has mask input: [1]
const hasMaskInput = new ort.Tensor('float32', [0], [1]);

// Original image size: [2]
const origImSize = new ort.Tensor('float32',
  [imageHeight, imageWidth],
  [2]
);
```

#### 3. **Inference Execution**

```typescript
const feeds = {
  image_embeddings: embedding,  // Pre-computed or from encoder
  point_coords: pointCoords,
  point_labels: pointLabels,
  mask_input: maskInput,
  has_mask_input: hasMaskInput,
  orig_im_size: origImSize
};

const results = await model.run(feeds);
const maskTensor = results.masks; // [1, 1, 256, 256]
```

#### 4. **Mask Post-Processing** (from `maskUtils.tsx`)

```typescript
// Threshold mask at 0.0 (SAM convention)
const maskData = maskTensor.data as Float32Array;
const binaryMask = new Uint8Array(maskData.length);

for (let i = 0; i < maskData.length; i++) {
  binaryMask[i] = maskData[i] > 0.0 ? 255 : 0;
}

// Create PNG blob
const canvas = document.createElement('canvas');
canvas.width = 256;
canvas.height = 256;
const ctx = canvas.getContext('2d')!;

const imageData = ctx.createImageData(256, 256);
for (let i = 0; i < binaryMask.length; i++) {
  imageData.data[i * 4] = binaryMask[i];     // R
  imageData.data[i * 4 + 1] = binaryMask[i]; // G
  imageData.data[i * 4 + 2] = binaryMask[i]; // B
  imageData.data[i * 4 + 3] = 255;           // A
}
ctx.putImageData(imageData, 0, 0);

const blob = await new Promise<Blob>((resolve) => {
  canvas.toBlob(blob => resolve(blob!), 'image/png');
});
```

#### 5. **Integration with Annota** (leverage existing infrastructure)

```typescript
// Convert blob to URL
const blobUrl = URL.createObjectURL(blob);

// Use existing Annota mask loader with binary type
const annotations = await loadMaskPolygons(blobUrl, {
  maskType: 'binary'  // Use the new binary mask support from v0.7.13
});

// Clean up
URL.revokeObjectURL(blobUrl);

// Add to annotator
annotator.state.store.add(annotations[0]);
```

### ONNX Model Export

Users need to export SAM models to ONNX format. Example Python script:

```python
import torch
from segment_anything import sam_model_registry
from segment_anything.utils.onnx import SamONNXModel

# Load SAM checkpoint
sam = sam_model_registry["vit_h"](checkpoint="sam_vit_h_4b8939.pth")

# Create ONNX model wrapper (decoder only)
onnx_model = SamONNXModel(sam, return_single_mask=True)

# Define inputs
embed_dim = sam.prompt_encoder.embed_dim
embed_size = sam.prompt_encoder.image_embedding_size
mask_input_size = [4 * x for x in embed_size]

dummy_inputs = {
    "image_embeddings": torch.randn(1, embed_dim, *embed_size, dtype=torch.float),
    "point_coords": torch.randint(low=0, high=1024, size=(1, 5, 2), dtype=torch.float),
    "point_labels": torch.randint(low=0, high=4, size=(1, 5), dtype=torch.float),
    "mask_input": torch.randn(1, 1, *mask_input_size, dtype=torch.float),
    "has_mask_input": torch.tensor([1], dtype=torch.float),
    "orig_im_size": torch.tensor([1500, 2250], dtype=torch.float),
}

# Export to ONNX
torch.onnx.export(
    onnx_model,
    tuple(dummy_inputs.values()),
    "sam_vit_h_decoder.onnx",
    export_params=True,
    opset_version=17,
    do_constant_folding=True,
    input_names=list(dummy_inputs.keys()),
    output_names=["masks", "iou_predictions", "low_res_masks"],
    dynamic_axes={
        "point_coords": {1: "num_points"},
        "point_labels": {1: "num_points"},
    }
)
```

For encoder export, see: https://github.com/facebookresearch/segment-anything/blob/main/notebooks/onnx_model_example.ipynb

### Tool: `SamTool`

```typescript
import { SamTool } from 'annota/tools';

const samTool = new SamTool(annotator, {
  // Model configuration
  decoderModelUrl: '/models/sam_vit_h_decoder.onnx',

  // Embedding options (choose one)
  embeddingMode: 'precomputed', // or 'realtime'
  embeddingUrl: '/embeddings/image123.npy', // For precomputed mode
  encoderModelUrl: '/models/sam_vit_h_encoder.onnx', // For realtime mode

  // Visualization
  showHoverPreview: true,
  previewOpacity: 0.5,

  // Annotation properties
  annotationProperties: {
    classification: 'positive',
    source: 'sam-onnx'
  }
});

// Activate tool
annotator.setTool(samTool);
```

### Helper Functions

```typescript
// Utility to create dummy/test embeddings
import { createDummyEmbedding } from 'annota/utils';
const dummyEmbedding = createDummyEmbedding(); // [1, 256, 64, 64]

// Load precomputed embedding from .npy file
import { loadNpyEmbedding } from 'annota/utils';
const embedding = await loadNpyEmbedding('/embeddings/image.npy');

// Future: Real-time encoding
import { encodeImage } from 'annota/utils';
const embedding = await encodeImage(imageUrl, encoderSession);
```

### Phase 1: Core Infrastructure

- [ ] Add `onnxruntime-web` dependency
- [ ] Create `SamOnnxModel` class for model management
  - [ ] Model initialization and caching
  - [ ] Inference method with click coordinates
  - [ ] Mask tensor → PNG blob conversion
- [ ] Create `SamTool` that extends `BaseTool`
  - [ ] Tool activation/deactivation lifecycle
  - [ ] Mouse hover preview (highlight detected object)
  - [ ] Click handler to create annotation
  - [ ] Loading states and error handling
- [ ] Integrate with existing `loadMaskPolygons()`
  - [ ] Already supports `maskType: 'binary'` as of v0.7.13
  - [ ] Leverages OpenCV.js for contour detection

### Phase 2: Embedding Management

- [ ] Add `npyjs` dependency for `.npy` file loading
- [ ] Create embedding loader utilities
  - [ ] `loadNpyEmbedding()` for precomputed files
  - [ ] `createDummyEmbedding()` for testing
- [ ] Add embedding caching/storage
  - [ ] IndexedDB for browser persistence
  - [ ] LRU cache for memory management

### Phase 3: UI/UX Polish

- [ ] Add model download progress indicator
- [ ] Show hover preview overlay
- [ ] Add confidence score display
- [ ] Keyboard shortcuts (Enter to confirm, Esc to cancel)
- [ ] Error messages for missing models/embeddings

### Phase 4: Performance Optimization

- [ ] Web Worker support for inference (avoid blocking main thread)
- [ ] SharedArrayBuffer for multi-threading (when available)
- [ ] Model quantization support for smaller file sizes
- [ ] WebGPU backend support (future ONNX Runtime feature)

### Phase 5: Advanced Features (Future)

- [ ] Multi-point prompts (positive + negative clicks)
- [ ] Box prompts (drag to select region)
- [ ] Iterative refinement (use previous mask as input)
- [ ] Real-time encoder support for dynamic images
- [ ] Support for MobileSAM (faster, smaller model)

## Performance Expectations

**Phase 1 (Decoder-only with precomputed embeddings):**
- Model download: ~10MB one-time
- Initialization: ~1-2 seconds
- Per-click inference: ~100-300ms
- Memory: ~500MB

**Phase 2 (With real-time encoder):**
- Model download: ~360MB one-time
- Initialization: ~5-10 seconds
- Image encoding: ~2-5 seconds
- Per-click inference: ~100-300ms
- Memory: ~2GB

## Browser Compatibility

**Minimum Requirements:**
- WebAssembly support (all modern browsers)
- ES2020 features
- ~2GB RAM for full model

**Recommended:**
- Chrome/Edge 90+ (best WASM performance)
- Firefox 89+ (good WASM support)
- Safari 15+ (WebAssembly, but slightly slower)

## Success Criteria

1. ✅ User can click on any object and get instant polygon annotation
2. ✅ No backend/API required - fully client-side
3. ✅ Works offline after initial model download
4. ✅ Smooth hover preview (no lag)
5. ✅ Performance: <300ms per click on modern hardware
6. ✅ Clean integration with existing Annota API

## References

- **SAM Paper**: https://arxiv.org/abs/2304.02643
- **SAM GitHub**: https://github.com/facebookresearch/segment-anything
- **SAM Demo**: https://github.com/facebookresearch/segment-anything/tree/main/demo
- **ONNX Runtime Web**: https://onnxruntime.ai/docs/tutorials/web/
- **Annota v0.7.13**: Binary mask support via `maskType: 'binary'` option

---

**Additional Context**: This builds on the recent v0.7.13 enhancement where `loadMaskPolygons()` gained native binary mask support. The Facebook SAM demo has proven browser-based SAM inference is feasible and provides excellent UX. By bringing this into Annota, we can offer a best-in-class annotation tool that works entirely offline and respects user privacy.
